File: sdks/jvm/src/main/kotlin/com/heossi/qnsi/CryptoInventoryClient.kt
package com.heossi.qnsi

import com.heossi.qnsi.internal.Internal
import com.heossi.qnsi.internal.RequestOptions
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonObject

private const val PREFIX = "/proxy/crypto/v1"

/** Kick off a discovery run over the given targets/modes. */
@Serializable
public data class DiscoverAssetsRequest @JvmOverloads constructor(
    public val targets: List<String>? = null,
    public val modes: List<String>? = null,
    public val options: JsonObject? = null,
)

/** Cryptographic asset inventory / CBOM. Wraps crypto-inventory-service `/crypto/v1`. */
public class CryptoInventoryClient internal constructor(private val transport: Internal) {
    @JvmOverloads
    public fun listAssets(query: Map<String, String?> = emptyMap()): JsonObject =
        transport.requestJson("GET", "$PREFIX/assets", null, RequestOptions(query = query))

    public fun getAsset(assetId: String): JsonObject =
        transport.requestJson("GET", "$PREFIX/assets/$assetId", null)

    public fun getAssetStats(tenantId: String): JsonObject =
        transport.requestJson("GET", "$PREFIX/assets/stats", null)

    @JvmOverloads
    public fun discoverAssets(req: DiscoverAssetsRequest, idempotencyKey: String? = null): JsonObject =
        transport.requestJson(
            "POST", "$PREFIX/assets/discover",
            transport.json.encodeToString(DiscoverAssetsRequest.serializer(), req),
            RequestOptions(idempotencyKey = idempotencyKey),
        )

    public fun getReadinessScore(tenantId: String): JsonObject =
        transport.requestJson("GET", "$PREFIX/readiness", null)
}

File: sdks/swift/Sources/QNSI/Clients/CryptoInventoryClient.swift
import Foundation

private let prefix = "/proxy/crypto/v1"

/// Cryptographic asset inventory / CBOM. Wraps crypto-inventory-service `/crypto/v1`.
public struct CryptoInventoryClient: Sendable {
    let transport: Transport

    public func listAssets(query: [String: String?] = [:]) async throws -> JSONValue {
        try await transport.requestJson("GET", "\(prefix)/assets", options: RequestOptions(query: query))
    }

    public func getAsset(assetId: String) async throws -> JSONValue {
        try await transport.requestJson("GET", "\(prefix)/assets/\(assetId)")
    }

    /// Asset stats for the activated tenant (tenantId is injected into the query).
    public func getAssetStats() async throws -> JSONValue {
        try await transport.requestJson("GET", "\(prefix)/assets/stats")
    }

    /// Kick off a discovery run over the given targets/modes.
    public func discoverAssets(
        targets: [String]? = nil,
        modes: [String]? = nil,
        options: [String: JSONValue]? = nil,
        idempotencyKey: String? = nil
    ) async throws -> JSONValue {
        var object: [String: JSONValue] = [:]
        if let targets { object["targets"] = .array(targets.map { .string($0) }) }
        if let modes { object["modes"] = .array(modes.map { .string($0) }) }
        if let options { object["options"] = .object(options) }
        return try await transport.requestJson(
            "POST", "\(prefix)/assets/discover", body: .object(object),
            options: RequestOptions(idempotencyKey: idempotencyKey)
        )
    }

    /// PQC readiness score for the activated tenant (tenantId is injected into the query).
    public func getReadinessScore() async throws -> JSONValue {
        try await transport.requestJson("GET", "\(prefix)/readiness")
    }
}
